### PR TITLE
Add updateSnap function to SnapController

### PR DESCRIPTION
**Pull request by ritave**
First PR for #223

### CODE REVIEW

File: packages/controllers/jest.config.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ module.exports = {
   coverageThreshold: {
     global: {
       branches: 57,
-      functions: 77,
+      functions: 76,
```
Comment (FrederikBolding) [inline packages/controllers/jest.config.js:12]: I don't think we should just arbitrarily reduce this, why is this needed?

File: packages/controllers/src/permissions/PermissionController.ts [hunk 2]
```diff
@@ -218,6 +218,14 @@ export type RequestPermissions = {
   handler: GenericPermissionController['requestPermissions'];
 };
 
+/**
+ * Removes for each origin removes given permissions
+ */
+export type RevokePermissions = {
```
Comment (FrederikBolding) [inline packages/controllers/src/permissions/PermissionController.ts:224]: This should be added in https://github.com/MetaMask/controllers since this version will be deleted as soon as the new controllers version is released.

File: packages/controllers/src/snaps/SnapController.ts [hunk 3]
```diff
@@ -329,7 +342,7 @@ type SnapControllerArgs = {
 
 type AddSnapBase = {
   id: SnapId;
-  version?: string;
+  versionRange?: string;
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:345]: Is there a reason for the rename? Do you think just using `version` is confusing? 🤔

File: packages/controllers/src/utils.ts [hunk 4]
```diff
@@ -52,6 +52,71 @@ export function isValidJson(value: unknown): value is Json {
   }
 }
 
+export function objectDiff<A, B>(a: A, b: B): Diff<A, B> {
+  return Object.entries(a).reduce((acc, [key, value]) => {
+    if (key in b) {
+      return acc;
+    }
+    acc[key] = value;
+    return acc;
```
Comment (FrederikBolding) [inline packages/controllers/src/utils.ts:61]: All of this could be simplified to a simple arrow function
```
=> key in b ? acc : ({ ...acc, [key]: value })
```

File: packages/controllers/src/snaps/SnapController.ts [hunk 5]
```diff
@@ -1471,3 +1556,11 @@ function isValidSnapVersionRange(version: unknown): version is string {
     typeof version === 'string' && validRangeSemver(version) !== null,
   );
 }
+
+function isLocalSnap(snapId: string): boolean {
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1560]: Would it make more sense to return an enum and get the snap type? So detecting if it's either local or npm in one function. Is easier to expand in the future.

File: packages/controllers/src/utils.ts [hunk 6]
```diff
@@ -52,6 +52,71 @@ export function isValidJson(value: unknown): value is Json {
...
+  }, {} as any);
```
Comment (FrederikBolding) [inline packages/controllers/src/utils.ts:62]: Is the cast needed here? Or can we use `reduce` generically?

File: packages/controllers/src/utils.ts [hunk 7]
```diff
@@ -52,6 +52,71 @@ export function isValidJson(value: unknown): value is Json {
...
+ * ```
+ * type t1 = Diff<{a: string, b: string}, {a: number}>
+ * // t1 = {b: string};
+ * type t2 = Diff<{a: string, 0: string}, Record<string, unknown>>;
+ * // t2 = { a?: string, 0: string};
+ * type t3 = Diff<{a: string, 0: string, 1: string}, Record<1 | string, unknown>>;
+ * // t3 = {a?: string, 0: string}
+ * ```
+ */
+export type Diff<A, B> = Omit<A, _LiteralKeys<B>> &
```
Comment (FrederikBolding) [inline packages/controllers/src/utils.ts:117]: @rekmarks would appreciate some help reviewing this part 😅

File: packages/controllers/src/snaps/SnapController.ts [hunk 8]
```diff
@@ -1039,6 +1052,56 @@ export class SnapController extends BaseController<
...
+      id: snapId as ValidatedSnapId,
+      manifest: newSnap.manifest,
+      sourceCode: newSnap.sourceCode,
+      versionRange: newVersionRange,
+    });
+
+    await this.authorize(snapId);
+
+    if (wasRunning) {
+      await this._startSnap({ snapId, sourceCode: newSnap.sourceCode });
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1097]: What happens to requests to the snap meanwhile? Is this part of https://github.com/MetaMask/snaps-skunkworks/issues/245 🤔

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: packages/controllers/src/permissions/PermissionController.ts (hunk 2 above)
Comment (FrederikBolding) [reply packages/controllers/src/permissions/PermissionController.ts:224]: @rekmarks is planning to release the new version today, so you might be able to get these changes in before then

Comment (ritave) [reply packages/controllers/src/permissions/PermissionController.ts:224]: Done https://github.com/MetaMask/controllers/pull/708

File: packages/controllers/src/snaps/SnapController.ts (hunk 3 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:345]: Variables in SnapController named `version` are used to handle a *concrete* version of a snap. Range on the other hand, can handle any SemVer compatible range. This was renamed to better represent the functionality of add() family functions

File: packages/controllers/src/utils.ts (hunk 4 above)
Comment (ritave) [reply packages/controllers/src/utils.ts:61]: Clever, but this creates a new object a has to iterate on all keys on every iteration, making it into O(n^2) algorithm. Compared to reused single object in the current one with O(n) complexity.

Comment (FrederikBolding) [reply packages/controllers/src/utils.ts:61]: I guess it depends on if we value readability over performance. Not sure what our standards are for that kind of thing. 

Either way if you want to keep this I would simplify to
```
if(!(key in b)){
  acc[key] = value;
}
return acc;
```

File: packages/controllers/src/snaps/SnapController.ts [hunk 9]
```diff
@@ -1039,6 +1052,56 @@ export class SnapController extends BaseController<
...
+   * @param newVersionRange A semver version range in which the maximum version will be chosen
+   * @returns @type {TruncatedSnap} if updated, @type {null} otherwise
+   */
+  async updateSnap(
+    snapId: SnapId,
+    newVersionRange: string = DEFAULT_REQUESTED_SNAP_VERSION,
+  ): Promise<TruncatedSnap | null> {
+    const snap = this.get(snapId);
+    if (snap === undefined) {
+      throw new Error(`Couldn't find snap ${snapId} to update`);
```
Comment (rekmarks) [inline packages/controllers/src/snaps/SnapController.ts:1067]:
```suggestion
      throw new Error(`Could not find snap ${snapId}. Install the snap before attempting to update it.`);
```

**Review by rekmarks — CHANGES_REQUESTED** [review]
This is an excellent start! All my comments are inline, but I'd like to call out my suggestion that we add an `updating` state to the snap state machine. We could potentially reuse the existing `installing` state, but we may want to differentiate between the two in various ways.

File: packages/controllers/src/utils.ts [hunk 10]
```diff
@@ -52,6 +52,71 @@ export function isValidJson(value: unknown): value is Json {
   }
 }
 
+export function objectDiff<A, B>(a: A, b: B): Diff<A, B> {
```
[the remark this answers is not in the document]
Comment (ritave) [reply packages/controllers/src/utils.ts:55]: 1. Sure
2. Sure
3. Don't think it's necessary - no need to add additional branches for  a case that's almost won't ever happend and is handled by the normal flow.
4. This is an expected behaviour. The function does [Set Difference](https://proofwiki.org/wiki/Definition:Set_Difference#:~:text=The%20(set)%20difference%20between%20two,%E2%88%88S%3Ax%E2%88%89T%7D). I'll change its name to better illustrate the behaviour.

File: packages/controllers/src/utils.ts (hunk 6 above)
Comment (ritave) [reply packages/controllers/src/utils.ts:62]: The cast is necessary since it's `{}` is not castable to `Diff<A,B>` as it lacks parameters of A.

File: packages/controllers/src/utils.ts [hunk 11]
```diff
@@ -52,6 +52,71 @@ export function isValidJson(value: unknown): value is Json {
...
+ * // t1 = {b: string};
+ * type t2 = Diff<{a: string, 0: string}, Record<string, unknown>>;
+ * // t2 = { a?: string, 0: string};
+ * type t3 = Diff<{a: string, 0: string, 1: string}, Record<1 | string, unknown>>;
+ * // t3 = {a?: string, 0: string}
+ * ```
+ */
+export type Diff<A, B> = Omit<A, _LiteralKeys<B>> &
+  Partial<Pick<A, Extract<keyof A, _NonLiteralKeys<B>>>>;
+
```
[the remark this answers is not in the document]
Comment (ritave) [reply packages/controllers/src/utils.ts:119]: The trick here is that the object you provided to `bar` is downcasted into a subset of it that matches `t2`.

It will fail when you actually try to use it:
```typescript
type t2 = Diff<{a: number}, {a: string, b: string}>
const bar: t2 = { b: 'foo', a: 'x', c: 'y' } // no error
const foo = bar.a // Property 'a' does not exist on type `t2`
```

I could constraint it using [`never` types](https://stackoverflow.com/a/44425486), but I think it would make working with Diff objects harder.

File: packages/controllers/src/snaps/SnapController.ts [hunk 12]
```diff
@@ -403,6 +416,7 @@ const snapStatusStateMachineConfig = {
           target: SnapStatus.running,
           cond: disabledGuard,
         },
+        [SnapStatusEvent.update]: SnapStatus.installing,
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:419]: Would it be better to have a `SnapStatus.updating`? I wonder if there will be a need to distinguish between a first-time install or an update 🤔

File: packages/controllers/src/snaps/SnapController.ts [hunk 13]
```diff
@@ -1054,6 +1068,62 @@ export class SnapController extends BaseController<
...
+      );
+    }
+
+    const newSnap = await this._fetchSnap(
+      snapId as ValidatedSnapId,
+      newVersionRange,
+    );
+    if (!gtSemver(newSnap.manifest.version, snap.version)) {
+      console.warn(
+        `Tried updating snap ${snapId} within ${newVersionRange} version range, but newer version ${snap.version} already set-up`,
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1095]: Nit: "... already _installed_"?

File: packages/controllers/src/snaps/SnapController.ts [hunk 14]
```diff
@@ -1054,6 +1068,62 @@ export class SnapController extends BaseController<
...
+  /**
+   * Updates, re-authorizes and then restarts given snap.
+   *
+   * @param snapId The id of the Snap to be updated
+   * @param newVersionRange A semver version range in which the maximum version will be chosen
+   * @returns @type {TruncatedSnap} if updated, @type {null} otherwise
+   */
+  async updateSnap(
+    snapId: SnapId,
+    newVersionRange: string = DEFAULT_REQUESTED_SNAP_VERSION,
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1080]: Do we need to use `resolveVersion` somewhere here for backwards compatibility with `latest`?

File: packages/controllers/src/utils.ts (hunk 6 above)
Comment (FrederikBolding) [reply packages/controllers/src/utils.ts:79]: Hmm. @Mrtenz is there a way to type this reduce proper or would you also cast?

File: packages/controllers/src/snaps/SnapController.ts [hunk 15]
```diff
@@ -1251,18 +1321,19 @@ export class SnapController extends BaseController<
...
+        case SnapIdPrefixes.local:
+          return this._fetchLocalSnap(snapId.replace(SnapIdPrefixes.local, ''));
+        case SnapIdPrefixes.npm:
+          return this._fetchNpmSnap(
+            snapId.replace(SnapIdPrefixes.npm, ''),
+            version,
+          );
+        default:
+          // This whill fail to compile if the above switch is not fully exhaustive
+          return assertExhaustive(snapPrefix);
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1335]: Should we just throw `Invalid snap id` or something similar here instead?

Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:1335]: (as we did before below)

File: packages/controllers/src/snaps/SnapController.test.ts [hunk 16]
```diff
@@ -835,6 +836,71 @@ describe('SnapController', () => {
...
+      .mockImplementationOnce(() => true);
+
+    await controller.installSnaps(origin, {
+      [FAKE_SNAP_ID]: { version: '>0.9.0 <1.1.0' },
+    });
+
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const newSnap = controller.get(FAKE_SNAP_ID)!;
+
+    // Notice usage of toBe - we're checking if it's actually the same object, not an equal one
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.test.ts:878]: Hmm, why is that important in this context?

File: packages/controllers/src/snaps/SnapController.ts (hunk 12 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:419]: Specifically decided not have additional state. Don't think it's necessary and it complicates logic

File: packages/controllers/src/snaps/SnapController.ts (hunk 15 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:1335]: Doesn't matter, won't reach it anyways

File: packages/controllers/src/snaps/SnapController.test.ts (hunk 16 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.test.ts:878]: To ensure that we don't actually manipulate the snap in any way

File: packages/controllers/src/snaps/SnapController.ts (hunk 14 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:1080]: Seems `resolveVersion` is only used when getting direct input from JSON-RPC, and `DEFAULT_REQUESTED_SNAP_VERSION` is used otherwise. This function is not directly called from RPC

File: packages/controllers/src/snaps/SnapController.ts (hunk 15 above)
Comment (rekmarks) [reply packages/controllers/src/snaps/SnapController.ts:1335]: Yeah, TypeScript should guarantee that we'll fail at compile-time before that ever happens.

File: packages/controllers/src/snaps/SnapController.ts (hunk 12 above)
Comment (rekmarks) [reply packages/controllers/src/snaps/SnapController.ts:419]: I've thought about whether we should have `SnapStatus.updating`, but we _could_ be fine without it, and it's quite nice to not have to worry about it in this PR. We can always complicate the state machine later if needed!

File: packages/controllers/src/snaps/SnapController.ts [hunk 17]
```diff
@@ -1364,14 +1435,40 @@ export class SnapController extends BaseController<
...
+      );
+
+      const newPermissions = setDiff(
         initialPermissions,
+        alreadyApprovedPermissions ?? {},
       );
-      return Object.values(approvedPermissions).map(
-        (perm) => perm.parentCapability,
+
+      const unusedPermissions = Object.keys(
```
Comment (rekmarks) [inline packages/controllers/src/snaps/SnapController.ts:1450]: Lest we forget:
```suggestion
      // TODO: The assumption that these are unused only holds so long as we do not
      // permit dynamic permission requests.
      const unusedPermissions = Object.keys(
```

File: packages/controllers/src/snaps/SnapController.ts [hunk 18]
```diff
@@ -1054,6 +1068,62 @@ export class SnapController extends BaseController<
...
+    newVersionRange: string = DEFAULT_REQUESTED_SNAP_VERSION,
+  ): Promise<TruncatedSnap | null> {
+    const snap = this.get(snapId);
+    if (snap === undefined) {
+      throw new Error(
+        `Could not find snap ${snapId}. Install the snap before attempting to update it.`,
+      );
+    }
+
+    const newSnap = await this._fetchSnap(
```
Comment (rekmarks) [inline packages/controllers/src/snaps/SnapController.ts:1089]: Although the logic of this function is sound either way, we should check whether the the requested version _range_ is valid before fetching the snap. Fetching a snap may be expensive, and I think this optimization would be worth it. I'd say that we could do this in the RPC method implementation, but we may want to call this method from within MetaMask itself, in which case we'd have to implement that check in multiple places.

File: packages/controllers/src/snaps/SnapController.ts (hunk 14 above)
Comment (rekmarks) [reply packages/controllers/src/snaps/SnapController.ts:1080]: Yeah, that seems reasonable to me.

**Review by rekmarks — CHANGES_REQUESTED** [review]
Alright, this is very close to being done! Just a handful of minor things.

The types in utils and the `setDiff` implementation are indeed really nice.

File: packages/controllers/src/snaps/SnapController.ts (hunk 17 above)
Comment (FrederikBolding) [reply packages/controllers/src/snaps/SnapController.ts:1450]: Good point!

File: packages/controllers/src/snaps/SnapController.ts [hunk 19]
```diff
@@ -1496,3 +1593,12 @@ function isValidSnapVersionRange(version: unknown): version is string {
     typeof version === 'string' && validRangeSemver(version) !== null,
   );
 }
+
+function snapIdToSnapPrefix(snapId: string): SnapIdPrefixes {
+  for (const prefix of Object.values(SnapIdPrefixes)) {
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/SnapController.ts:1598]: Nit: Could be simplified to `const value = Object.values(SnapIdPrefixes).find(prefix => snapId.startsWith(prefix));`?

**Review by FrederikBolding — COMMENTED** [review]
One small comment, but overall once @rekmarks comments are addressed I think this is good to go!

File: packages/controllers/src/utils.ts (hunk 6 above)
Comment (Mrtenz) [reply packages/controllers/src/utils.ts:79]: @ritave @FrederikBolding I think in this case it might be better to do something like this:
```ts
return Object.entries(objectA).reduce<Record<string, unknown>>(
  (acc, [key, value]) => {
    if (!(key in objectB)) {
      acc[key] = value;
    }
    return acc;
  },
  {},
) as Diff<ObjectA, ObjectB>;
```

This keeps the reducer strictly typed, and only the result of that is casted to `Diff`.

File: packages/controllers/src/snaps/SnapController.ts (hunk 18 above)
Comment (ritave) [reply packages/controllers/src/snaps/SnapController.ts:1089]: It's already checked in `_fetchNpmSnap` but I'll add a check anyways

**Review by FrederikBolding — DISMISSED** [review]
good to merge once @rekmarks comments are addressed!

**Review by rekmarks — DISMISSED** [review]
Ship it!

**Review by FrederikBolding — APPROVED** [review]

**Review by FrederikBolding — APPROVED** [review]
